### PR TITLE
Introduce sidebar with file browser

### PR DESCRIPTION
**Pull request by christopher-l**
Finally got back to it. Changes compared to last time are:
- Adapted the design to closer match Gnome Builder:
    - No expander arrows.
    - Directories are expanded, not opened, with single clicks.
- Swapped the toolbar for a ComboBox with the components of the cwd path.
- Sidebar is hidden by default.
- Visibility state and width are restored at startup.

### CODE REVIEW

**Comment by wezm** [comment]
This is wonderful. Adding a screenshot (GNOME 3.16 on FreeBSD):

screenshot from 2018-03-12 13-38-03

**Comment by daa84** [comment]
Looks cool 👍 
Need some time to check and read code 😄

File: src/file_browser.rs [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+}
+
+/// Updates the dirctory list on top of the file browser.
+///
+/// The list represents the path the the current working directory.  If the new cwd is a parent of
+/// the old one, the list is kept and only the active entry is updated. Otherwise, the list is
+/// replaced with the new path and the last entry is marked active.
+fn update_dir_list(dir: &str, dir_list_model: &gtk::TreeStore, dir_list: &gtk::ComboBox) {
+    // The current working directory path.
+    let complete_path = Path::new(dir).canonicalize().unwrap();
```
Comment (daa84) [inline src/file_browser.rs]: Get crash here on Windows when select "/" as parent directory. Not sure what the best solution here.
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 1, message: "Incorrect function." } }', src\libcore\result.rs:916:5
```
Maybe "/" is just not allowed for windows as directory, so it is need to remove it from drop down list?

File: src/file_browser.rs [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+/// Populates one level, i.e. one directory of the file browser tree.
+fn populate_tree_nodes(
+    store: &gtk::TreeStore,
+    state: &State,
+    dir: &str,
+    parent: Option<&gtk::TreeIter>,
+) {
+    let path = Path::new(dir);
+    let iter = path.read_dir()
+        .expect("read dir failed")
```
Comment (daa84) [inline src/file_browser.rs]: Here and can be in some other places, call `expect` is not good i think. As error can of reading directory can crush all application.

Comment (christopher-l) [reply src/file_browser.rs]: Oh, forgot about those. Removed two `expect()` calls from `file_browser.rs`. There are some `unwrap()` calls left, but as far as I can see, these shouldn't trigger unless there is an actual bug in the code.

**Comment by daa84** [comment]
Thank you :+1:

**Pull request merged**